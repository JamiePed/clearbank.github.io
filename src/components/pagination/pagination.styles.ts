import styled from 'styled-components'

import { breakpoints } from 'src/components/theme'

export const Container = styled.nav`
  display: flex;
  flex-wrap: wrap;
  justify-content: space-between;

  margin: 50px 0 40px 0;
  padding-top: 15px;

  @media screen and (min-width: ${breakpoints.medium}) {
    padding-top: 60px;
    flex-wrap: nowrap;
  }
`
