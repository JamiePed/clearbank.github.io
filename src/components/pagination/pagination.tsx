import React from 'react'

import { hasWindow } from 'src/utils/browser.services'
import PaginationLink from '../paginationLink/pagination-link'

import * as Types from './pagination.types'
import * as Styles from './pagination.styles'

const Pagination: React.FunctionComponent<Types.PaginationProps> = ({
  menuItems
}): JSX.Element => {
  const currentPath = hasWindow() ? window.location.pathname : ''
  const currentSlug = currentPath.replace(/\/$/, '')

  const currentMenuIndex = menuItems.findIndex((item: any) => {
    const menuSlug: string = item.menuItem.slug

    return menuSlug.toUpperCase() === currentSlug.toUpperCase()
  })

  if (currentMenuIndex === -1) {
    return null
  }

  let prevPage = null
  let nextPage = null

  // has prev page
  if (typeof menuItems[currentMenuIndex - 1] !== 'undefined') {
    prevPage = menuItems[currentMenuIndex - 1].menuItem
  }

  // has next page
  if (typeof menuItems[currentMenuIndex + 1] !== 'undefined') {
    nextPage = menuItems[currentMenuIndex + 1].menuItem
  }

  return (
    <Styles.Container>
      {!!prevPage && (
        <PaginationLink
          direction={Types.direction.previous}
          key={prevPage.id}
          slug={prevPage.slug}
          title={prevPage.title}
        />
      )}
      {!!nextPage && (
        <PaginationLink
          direction={Types.direction.next}
          key={nextPage.id}
          slug={nextPage.slug}
          title={nextPage.title}
        />
      )}
    </Styles.Container>
  )
}

export default Pagination
