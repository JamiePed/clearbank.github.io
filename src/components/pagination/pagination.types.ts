export enum direction {
  previous = 'previous',
  next = 'next'
}

export interface PaginationProps {
  menuItems: any[]
}
