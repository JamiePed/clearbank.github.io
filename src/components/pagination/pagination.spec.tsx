import React from 'react'
import { render } from '@testing-library/react'
import '@testing-library/jest-dom'
import 'jest-styled-components'

import Component from './pagination'

import * as Types from './pagination.types'

let component: ReturnType<typeof render> = null
let fragment: DocumentFragment = null
let root: HTMLElement = null

jest.mock('src/utils/browser.services', () => ({
  hasWindow: () => {
    return true
  }
}))

const defaultProps: Types.PaginationProps = {
  menuItems: [
    {
      menuItem: {
        id: 'id1',
        order: 1,
        slug: '/doc/slug1',
        title: 'title1'
      },
      subMenuItems: [
        {
          node: {
            fields: {
              id: 'subid1',
              order: 1,
              slug: '/doc/slug1/subslug1',
              title: 'subtitle1'
            }
          }
        },
        {
          node: {
            fields: {
              id: 'subid2',
              order: 2,
              slug: '/doc/slug1/subslug2',
              title: 'subtitle2'
            }
          }
        }
      ]
    },
    {
      menuItem: {
        id: 'id2',
        order: 2,
        slug: '/doc/slug2',
        title: 'title2'
      },
      subMenuItems: [
        {
          node: {
            fields: {
              id: 'subid1',
              order: 1,
              slug: '/doc/slug2/subslug1',
              title: 'subtitle1'
            }
          }
        },
        {
          node: {
            fields: {
              id: 'subid2',
              order: 2,
              slug: '/doc/slug2/subslug2',
              title: 'subtitle2'
            }
          }
        }
      ]
    },
    {
      menuItem: {
        id: 'id3',
        order: 3,
        slug: '/doc/slug3',
        title: 'title3'
      },
      subMenuItems: [
        {
          node: {
            fields: {
              id: 'subid1',
              order: 1,
              slug: '/doc/slug3/subslug1',
              title: 'subtitle1'
            }
          }
        },
        {
          node: {
            fields: {
              id: 'subid2',
              order: 2,
              slug: '/doc/slug3/subslug2',
              title: 'subtitle2'
            }
          }
        }
      ]
    }
  ]
}

beforeAll(() => {
  // mock global location object
  delete window.location

  // has prev and next child
  window.location = { pathname: '/doc/slug2' }
})

beforeEach(() => {
  component = render(<Component {...defaultProps} children='children' />)

  fragment = component.asFragment()
  root = component.container.firstChild
})

afterEach(() => {
  fragment = null
  root = null
})

describe('Component', () => {
  test('matches snapshot', () => {
    expect(root).toMatchSnapshot()
  })

  test('is a nav element', () => {
    expect(fragment.querySelectorAll('nav').length).toBe(1)
    expect(root.tagName.toLowerCase()).toBe('nav')
  })

  test('is not empty', () => {
    expect(root.textContent.length).toBeTruthy()
  })

  test('has no menu when menuList is empty', () => {
    component.rerender(<Component menuItems={[]} />)

    expect(component.asFragment().hasChildNodes()).toBeFalsy()
  })

  test('has no menu when slug is not found in menuList', () => {
    window.location = { pathname: '/path/subpath' }
    component.rerender(<Component menuItems={[]} />)

    expect(component.asFragment().hasChildNodes()).toBeFalsy()

    // restore default state
    window.location = { pathname: '/doc/slug2' }
    component.rerender(<Component {...defaultProps} children='children' />)
  })

  test('has prev and next link', () => {
    expect(root.children.length).toBe(2)
  })

  test('has prev link before next link (correct order)', () => {
    expect(
      root.children[0].textContent.toLowerCase().includes('previous')
    ).toBeTruthy()

    expect(
      root.children[1].textContent.toLowerCase().includes('next')
    ).toBeTruthy()
  })

  test('has prev link with correct link, direction attribute and icon', () => {
    const link = root.children[0]

    expect(link.hasAttribute('direction')).toBeTruthy()
    expect(link.getAttribute('direction')).toBe('previous')

    expect(link.querySelectorAll('a').length).toBe(1)
    expect(link.querySelector('a').hasAttribute('href')).toBeTruthy()
    expect(link.querySelector('a').getAttribute('href')).toBe(
      defaultProps.menuItems[0].menuItem.slug
    )
    expect(link.querySelector('a').hasAttribute('direction')).toBeTruthy()
    expect(link.querySelector('a').getAttribute('direction')).toBe('previous')

    expect(link.querySelectorAll('.cb-icon-arrow-left').length).toBeTruthy
    expect(
      link.querySelector('.cb-icon-arrow-left').tagName.toLowerCase()
    ).toBe('i')
  })

  test('has prev link with correct link, direction attribute and icon', () => {
    const link = root.children[1]

    expect(link.hasAttribute('direction')).toBeTruthy()
    expect(link.getAttribute('direction')).toBe('next')

    expect(link.querySelectorAll('a').length).toBe(1)
    expect(link.querySelector('a').hasAttribute('href')).toBeTruthy()
    expect(link.querySelector('a').getAttribute('href')).toBe(
      defaultProps.menuItems[2].menuItem.slug
    )
    expect(link.querySelector('a').hasAttribute('direction')).toBeTruthy()
    expect(link.querySelector('a').getAttribute('direction')).toBe('next')

    expect(link.querySelectorAll('.cb-icon-arrow-right').length).toBeTruthy
    expect(
      link.querySelector('.cb-icon-arrow-right').tagName.toLowerCase()
    ).toBe('i')
  })

  test('has no prev link when slug is first in menu', () => {
    window.location = { pathname: '/doc/slug1' }
    component.rerender(<Component {...defaultProps} children='children' />)

    expect(root.children.length).toBe(1)
    expect(
      root.children[0].textContent.toLowerCase().includes('previous')
    ).toBeFalsy()
    expect(
      root.children[0].textContent.toLowerCase().includes('next')
    ).toBeTruthy()

    // restore default state
    window.location = { pathname: '/doc/slug2' }
    component.rerender(<Component {...defaultProps} children='children' />)
  })

  test('has no next link when slug is last in menu', () => {
    window.location = { pathname: '/doc/slug3' }
    component.rerender(<Component {...defaultProps} children='children' />)

    expect(root.children.length).toBe(1)
    expect(
      root.children[0].textContent.toLowerCase().includes('next')
    ).toBeFalsy()
    expect(
      root.children[0].textContent.toLowerCase().includes('previous')
    ).toBeTruthy()

    // restore default state
    window.location = { pathname: '/doc/slug2' }
    component.rerender(<Component {...defaultProps} children='children' />)
  })
})
