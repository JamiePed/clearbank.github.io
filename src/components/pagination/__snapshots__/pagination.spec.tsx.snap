// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component matches snapshot 1`] = `
.c2 {
  font-size: 14px;
  cursor: default;
  color: #6c7f98;
}

.c6 {
  margin-left: 10px;
  margin-right: 10px;
  font-size: 16px;
}

.c5 {
  width: 22px;
  height: 22px;
  font-size: 25px;
  -webkit-transform: translateX(0);
  -ms-transform: translateX(0);
  transform: translateX(0);
  will-change: transform;
  -webkit-transition: -webkit-transform 0.25s ease-out;
  -webkit-transition: transform 0.25s ease-out;
  transition: transform 0.25s ease-out;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: flex-end;
  -webkit-box-align: flex-end;
  -ms-flex-align: flex-end;
  align-items: flex-end;
  color: #000000;
  font-weight: bold;
  font-family: arial,sans-serif;
}

.c3:hover {
  color: #000000;
}

.c8 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: flex-end;
  -webkit-box-align: flex-end;
  -ms-flex-align: flex-end;
  align-items: flex-end;
  color: #000000;
  font-weight: bold;
  font-family: arial,sans-serif;
  -webkit-flex-direction: row-reverse;
  -ms-flex-direction: row-reverse;
  flex-direction: row-reverse;
}

.c8:hover {
  color: #000000;
}

.c1:hover .c4 {
  -webkit-transform: translateX(-5px);
  -ms-transform: translateX(-5px);
  transform: translateX(-5px);
}

.c7 {
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  text-align: right;
}

.c7:hover .c4 {
  -webkit-transform: translateX(5px);
  -ms-transform: translateX(5px);
  transform: translateX(5px);
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  margin: 50px 0 40px 0;
  padding-top: 15px;
}

@media screen and (min-width:768px) {
  .c0 {
    padding-top: 60px;
    -webkit-flex-wrap: nowrap;
    -ms-flex-wrap: nowrap;
    flex-wrap: nowrap;
  }
}

<nav
  class="c0"
>
  <div
    class="c1"
    direction="previous"
  >
    <span
      class="c2"
    >
      Previous
    </span>
    <a
      class="c3"
      direction="previous"
      href="/doc/slug1"
    >
      <span
        class="c4 c5"
      >
        <i
          class="cb-icon-arrow-left"
        />
      </span>
      <span
        class="c6"
      >
        title1
      </span>
    </a>
  </div>
  <div
    class="c7"
    direction="next"
  >
    <span
      class="c2"
    >
      Next
    </span>
    <a
      class="c8"
      direction="next"
      href="/doc/slug3"
    >
      <span
        class="c4 c5"
      >
        <i
          class="cb-icon-arrow-right"
        />
      </span>
      <span
        class="c6"
      >
        title3
      </span>
    </a>
  </div>
</nav>
`;
